### PR TITLE
feat: refactor shared structure 

### LINKED ISSUE #172: The configured `baseUrl` is not respected when resolving import paths
Thanks, @decmurphy for [reporting](https://github.com/NativeScript/nativescript-schematics/issues/117#issuecomment-448196154) the problem!

**Environment**
 - Angular: 7.1.0
 - Plugins: @nativescript/schematics 0.4.0

**Describe the bug**
The `ng add` command crashes with:
```
Could not find bootstrapped module.
```
when using custom `baseUrl` and absolute paths for the entry NgModule (`app.module`) or the entry component (`app.component`).

**Workaround**
Use relative imports for `app.module.ts` and `app.component.ts`.

**To Reproduce**
1. Create new Angular web project:
``` bash
ng new leapp
```

2. Configure the `baseUrl` in the `tsconfig.json`:
``` json
{
  "compilerOptions": {
    "baseUrl": "./src",
    ...
  }
}
```

3. Import `app.component` with absolute path in `app.module.ts`:
``` ts
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from 'app/app.component';

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

4. Try to add NativeScript to the project:
```
ng add @nativescript/schematics
```

**Current behavior**
We are not taking the `baseUrl` into account when resolving the import paths.

**Expected behaviour**
The code causing the problem is in the [`getEntryComponentMetadata`](https://github.com/NativeScript/nativescript-schematics/blob/master/src/angular-project-parser.ts#L181) function:
``` ts
  const importPath = findImportPath(source, className);


  const entryModuleDir = dirname(entryModulePath);
  const path = join(entryModuleDir, importPath) + '.ts';
```
The `importPath` for `AppComponent` is `app/app.component`. The `entryModuleDir` is `src/app`. The `join` function returns `src/app/app/app.component` which is obviously wrong and later causes the `Cannot find module` error.
The right behavior would be to detect that `app/app.component` is an absolute path and join it with the configured `baseUrl`.

**Sample project**
https://github.com/sis0k0/baseurl-ng-add-ns

Follows up #117.

### PR DESCRIPTION
**Pull request by vakrilov**
## The Problem
Imports for Web or Mobile specific files cause errors in IDEs and build time. 

Example: 
- You have service that has platform specific implementations for wen and mobile in `service.tns.ts` and `service.web.ts`. 
- The service is imported in another module with `import { xxx } from ".../service"`
- IDE (ex. VsCode) will (rightfully) complain that it cannot find the service module

## The Solution

Configure typescript to resolve platform specific files (with `.tns.ts`/`.web.ts` extension) with priority over shared files (just `*.ts`). The configuration is done by setting `paths` in `tsconfig`. Here is how the `tsconfig.tns.json` (used for mobile) looks:

```
"compilerOptions": {
   //...
   "paths": {
     "@src/*": [
       "src/*.android.ts",
       "src/*.ios.ts",
       "src/*.tns.ts",
       "src/*"
     ]
   }
 }
```

### Important Remarks

1. The common config contains mapping for both web and mobile so that paths can work and be resolved inside IDE (ex. VS Code) while developing.
2. Platform specific configs contain mapping only for the platform (see example above).
3. We map `@src/` to the `src` folder of the project. Other symbols be used - like `@proj/`. However, don’t use `“~/"` - this leads to confusing build-time errors and suffering. The reason: `“~/"` is reserved and is already used as an alias in webpack configuration it is needed for CSS files resolution.
4. The `“paths”` mapping only work for non-relative imports. This means that all imports that should be resolved to a plat-specific file should be defined with as non-relative paths starting with `@src`. Our suggestion is to adopt a convention to use `@src` for all project files imports and enforce it with a tslint rule.
5. Sometimes when building for NativeScript there is a problem with AOT compiler picking up `“*.ts”` extension that are intended for web-only. It then fails with a `component/module not used` error. To solve that we explicitly specify files in `tns` and `web` ts-configurations. These are the entry points to TS compilations. All lazy modules should be specified in `“files”` list in tsconfigs, as they are not directly imported in other TS files.

## Tasks

- [x] Modify `ng-new` schematic to create apps using the ts-configuration described above.
- [x] Create tslint rule that enforces non-relative imports. It should have an auto-fix option that replaces relative imports with the `@src` scheme
- [x] There are a couple of places in schematics that we need to resolve imports. Resolve them using the ts-configurations so that they can work with `@src` and `basePath` mappings
- [x] Make sure all schematics work and produce code with `@src` imports (instead of relative). We can just reuse the tslint rule. We can execute the rule with `--fix` to make it fix relative imports that we have generated
  - [x] `ng-new`
  - [x] `add-ns`
  - [x] `shared`
  - [x] `generate component`
  - [x] `generate module`
  - [x] `master-detail template`
  - [x] `migrate-module`
  - [x] `migrate-component`
- [ ] Update the documentation

Fixes #172

### CODE REVIEW

File: src/angular-project-parser.ts [hunk 1]
```diff
@@ -212,10 +218,83 @@ function getAppRootTag(tree: Tree, entryComponentPath: string): string {
...
+    const resolutionResult = ts.resolveModuleName(moduleName, containingFilePath, compilerOptions, moduleResolutionHost);
+    if (resolutionResult.resolvedModule) {
+      return resolutionResult.resolvedModule.resolvedFileName;
+    } else {
+      throw new SchematicsException(`Could not resolve ${moduleName}  using config: ${tsConfigName}`);
+    }
+  }
+}
+
+function tsResolve(moduleName: string, containingFilePath: string, tsConfigName: string, tree: Tree): string {
```
Comment (sis0k0) [inline src/angular-project-parser.ts]: This fn is unused. Do we need it?

Comment (vakrilov) [reply src/angular-project-parser.ts]: Nope

File: src/ng-new/shared/_files/__sourcedir__/tsconfig.app.json [hunk 2]
```diff
@@ -3,6 +3,16 @@
...
-  }
+    "types": [],
+    "paths": {
+      "@src/*": [
+        "<%= sourcedir %>/*.web",
+        "<%= sourcedir %>/*"
+      ]
+    }
+  },
+  "files": [
```
[the remark this answers is not in the document]
Comment (sis0k0) [reply src/ng-new/shared/_files/__sourcedir__/tsconfig.app.json]: Good catch! The file paths are relative to the `src` directory.
In Angular 7.0 the `tsconfig.app.json` file is in `my-project/src`. That's why the paths are `main.ts` and `polyfills.ts`. However, in Angular 8.0, the `tsconfig.app.json` file is on root level - in `./my-project`. The paths should be changed to `src/main.ts` and `src/polyfills.ts`.
This problem will be addressed in the PR, upgrading to Angular 8.0.

**Comment by sis0k0** [comment]
test

**Pull request merged**